File: java/com/commoncoder/learn/dsa/queue/PriorityQueue.java
package com.commoncoder.learn.dsa.queue;

public class PriorityQueue {

    private Node head = null;
    private int sizeVal = 0;


    // TC: O(N) SC: O(1)
    public void push(Value value) {
        sizeVal++;
        insertValueBasedOnHigherPriority(value);
    }

    // TC: O(1) SC: O(1)
    public Value poll() {
        if (isEmpty()) {
            throw new RuntimeException("PQ is empty!");
        }
        sizeVal--;
        return removeValueFromHead();
    }

    //TC: O(1) SC: O(1)
    public Value peek() {
        if (isEmpty()) {
            throw new RuntimeException("PQ is empty!");
        }
        return head.value;
    }

    public int size() {
        return sizeVal;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public static class Value {
        int data;
        int priority;

        public Value(int data, int priority) {
            this.data = data;
            this.priority = priority;
        }
    }

    private static class Node {
        Value value;
        Node next;

        public Node(Value value) {
            this.value = value;
        }

        public Node(Value value, Node next) {
            this.value = value;
            this.next = next;
        }
    }

    // TC: O(N) SC: O(1)
    private void insertValueBasedOnHigherPriority(Value newValue) {
        Node temp = head;
        Node prev = null;

        while (temp != null && temp.value.priority >= newValue.priority) {
            prev = temp;
            temp = temp.next;
        }
        Node newNode = new Node(newValue);
        newNode.next = temp;
        if (prev == null) {
            this.head = newNode;
        } else {
            prev.next = newNode;
        }

    }

    private Value removeValueFromHead() {
        if (head == null) {
            return null;
        }

        Value valueRemoved = head.value;
        this.head = head.next;
        return valueRemoved;
    }
}

File: cpp/29QueueBasics/priority_queue.cpp
#include <bits/stdc++.h>

using namespace std;

// This sample priority queue solution respects higher priority i.e.
// higher priority elements gets polled before lower priority elements.

// 10
//  3 3 3 2 [2]<---->[1] 1 1 1 1 1
class PriorityQueue {
 public:
  void push(int value, int priority) {
    sizeValue++;
    insertNewNodeAsPerPriority(value, priority);
  }

  int poll() {
    if (isEmpty()) {
      throw runtime_error("PQ is empty!");
    }
    sizeValue--;
    return *removeFirstNode();
  }

  int peek() {
    if (isEmpty()) {
      throw runtime_error("PQ is empty!");
    }
    return this->head->value;
  }

  int size() { return sizeValue; }

  bool isEmpty() { return head == nullptr; }

 private:
  class Node {
   public:
    int value;
    int priority;
    Node* next;

    Node(int value, int priority) {
      this->value = value;
      this->priority = priority;
      this->next = nullptr;
    }

    Node(int value, int priority, Node* next) {
      this->value = value;
      this->priority = priority;
      this->next = next;
    }
  };

  void insertNewNodeAsPerPriority(int value, int priority) {
    Node* newNode = new Node(value, priority);
    Node* prev = nullptr;
    Node* temp = head;
    while (temp != nullptr && temp->priority >= priority) {
      prev = temp;
      temp = temp->next;
    }
    newNode->next = temp;

    if (prev == nullptr) {
      this->head = newNode;
    } else {
      prev->next = newNode;
    }
  }

  int* removeFirstNode() {
    if (head == nullptr) {
      return nullptr;
    }
    Node* nodeToDelete = head;
    this->head = this->head->next;
    int valDeleted = nodeToDelete->value;
    delete nodeToDelete;
    return &valDeleted;
  }

  Node* head;
  int sizeValue = 0;
};